python/pyarrow/_dataset.pyx

```diff
@@ -2022,3 +2022,66 @@ def _get_partition_keys(Expression partition_expression):
         frombytes(name_val.first): pyarrow_wrap_scalar(name_val.second).as_py()
         for name_val in GetResultValue(CGetPartitionKeys(deref(expr.get())))
     }
+
+
+def _filesystemdataset_write(
+    data, object base_dir, Schema schema,
+    FileFormat format, FileSystem filesystem, Partitioning partitioning
+):
+    """
+    CFileSystemDataset.Write wrapper
+    """
+    cdef:
+        c_string c_base_dir
+        shared_ptr[CSchema] c_schema
+        shared_ptr[CFileFormat] c_format
+        shared_ptr[CFileSystem] c_filesystem
+        shared_ptr[CPartitioning] c_partitioning
+        shared_ptr[CScanContext] c_context
+        # to create iterator of InMemory fragments
+        vector[shared_ptr[CRecordBatch]] c_batches
+        shared_ptr[CFragment] c_fragment
+        vector[shared_ptr[CFragment]] c_fragment_vector
+
+    c_base_dir = tobytes(_stringify_path(base_dir))
+    c_schema = pyarrow_unwrap_schema(schema)
+    c_format = format.unwrap()
+    c_filesystem = filesystem.unwrap()
+    c_partitioning = partitioning.unwrap()
+    # passthrough use_threads?
+    c_context = _build_scan_context()
+
+    if isinstance(data, Dataset):
+        with nogil:
+            check_status(
+                CFileSystemDataset.Write(
+                    c_schema,
+                    c_format,
+                    c_filesystem,
+                    c_base_dir,
+                    c_partitioning,
+                    c_context,
+                    (<Dataset> data).dataset.GetFragments()
+                )
+            )
+    else:
+        # data is list of batches
+        for batch in data:
+            c_batches.push_back((<RecordBatch> batch).sp_batch)
+
+        c_fragment = shared_ptr[CFragment](
+            new CInMemoryFragment(c_batches, _true.unwrap()))
+        c_fragment_vector.push_back(c_fragment)
+
+        with nogil:
+            check_status(
+                CFileSystemDataset.Write(
+                    c_schema,
+                    c_format,
+                    c_filesystem,
+                    c_base_dir,
+                    c_partitioning,
+                    c_context,
+                    MakeVectorIterator(c_fragment_vector)
```

Comment:
Currently we only parallelize over written fragments, so since this only creates a single written fragment use_threads is effectively ignored. I'd recommend creating a fragment wrapping each batch. This will result in one or more files per input record batch however.
 ```suggestion
        # data is list of batches
        for batch in data:
            c_batches.push_back((<RecordBatch> batch).sp_batch)
            c_fragment = shared_ptr[CFragment](
                new CInMemoryFragment(c_batches, _true.unwrap()))
            c_batches.clear()
            c_fragment_vector.push_back(c_fragment)

        with nogil:
            check_status(
                CFileSystemDataset.Write(
                    c_schema,
                    c_format,
                    c_filesystem,
                    c_base_dir,
                    c_partitioning,
                    c_context,
                    MakeVectorIterator(move(c_fragment_vector))
```

Reply:
Yes, this is one of the aspects I wanted to ask: indeed right now by creating a single Fragment, it gets written to a single file. If you have a big table, you might want to make a file per batch. But on the other hand, that might also result in a lot of small files (certainly if you also split on some partition columns)? Eg in Parquet, you typically have files with multiple row groups (which might be somewhat comparable to multiple record batches)

So we should maybe have some configurability?

Reply:
@bkietz So I updated this to enable the use case of writing multiple fragments, but by explicit choice by the user -> passing a single Table/RecordBatch = writing a single fragment (even if the Table has multiple batches), passing a list of Table/RecordBatches -> written as multiple fragments. 

So if the user has a Table with multiple batches and wants to write this as multiple files instead of a single file (in case there is no partitioning), they can do `write_dataset(table.to_batches)`

Reply:
Great, this provides a minimal handle on write parallelism; this will be necessary until the c++ implementation can be made more robustly parallel than "one thread per written fragment".